### PR TITLE
Bug 2088634: fix(operator): adds validation after rendering dc to check for invali…

### LINKED ISSUE #459: [BUG] the catalog is not validated after package filtering
# Version

```console
$ oc-mirror version
The main branch on commit 7bde4c7bde6b829a34592b2d12115fb38b15576f
```

# What happened?

In use cases when ranges are defined on operator packages, it is possible to specify the versions in a way that breaks the package validity by breaking the upgrade graph or removing the default channel.

# What did you expect to happen?

`oc-mirror` to report an error to signify that the catalog is not valid during the dry-run.

# How to reproduce it (as minimally and precisely as possible)?

Mirror an operator package with a channel that is not the default channel.

# Anything else we need to know?

N/A

# References

N/A

### PR DESCRIPTION
**Pull request by jpower432**
…d index

Signed-off-by: Jennifer Power <barnabei.jennifer@gmail.com>

# Description

Using ranges to specify operator bundle versions can result in pruned
default channels and invalid replace chains (semver ordering is not
enforced on operator upgrade graphs). This adds a step that converts
the declarative config to a model which performs validation
on the FBC data.

Fixes #459

## Type of change

- [X] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

- [X] Unit tests added to check for errors for default channels and invalid graphs

# Checklist:

- [X] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: pkg/cli/mirror/operator.go [hunk 1]
```diff
@@ -705,3 +717,25 @@ func validate(dc *declcfg.DeclarativeConfig) error {
...
 }
+
+// checkError will check validation errors
+// from operator registry and return a modified
+// error messages for mirror usage
+// FIXME(jpower432): Checking against the errors string could be an issue since
+// this is depending on the strings returned from `opm validate`. It would be better to propose that
+// the validation error are exposed here https://github.com/operator-framework/operator-registry/blob/master/alpha/model/error.go
+// and adds structure errors that return package and channels information.
+func checkErr(err error) error {
```
Comment (jpower432) [inline pkg/cli/mirror/operator.go:728]: @dinhxuanvu I am kind of going back and forth on whether adding this is worth it given it depends on string values from a library. An ideal situation would be if the `validationErr` type in operator registry was exposed and maybe defined as a structure error. Do you think that is possible in the future?

File: pkg/cli/mirror/operator.go [hunk 2]
```diff
@@ -705,3 +717,29 @@ func validate(dc *declcfg.DeclarativeConfig) error {
...
+func (o *OperatorOptions) checkValidationErr(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	fmt.Fprintln(o.ErrOut, "\nThe rendered catalog is invalid.")
+	// handle known error causes
+	var validationMsg string
+	switch {
+	case strings.Contains(err.Error(), "channel must contain at least one bundle"):
```
Comment (dinhxuanvu) [inline pkg/cli/mirror/operator.go:737]: Hopefully one day we won't need to do this ugly string check for validation in FBC.

**Review by dinhxuanvu — COMMENTED** [review]
/lgtm

**Pull request merged**